clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -980,25 +1002,53 @@ protected ChangelogClientConfig getChangelogClientConfig() {
     return changelogClientConfig;
   }
 
-  private class HeartbeatReporterThread extends Thread {
-    HeartbeatReporterThread() {
+  protected HeartbeatReporterThread getHeartbeatReporterThread() {
+    return heartbeatReporterThread;
+  }
+
+  protected BasicConsumerStats getChangeCaptureStats() {
+    return changeCaptureStats;
+  }
+
+  protected class HeartbeatReporterThread extends Thread {
+    protected HeartbeatReporterThread() {
       super("Ingestion-Heartbeat-Reporter-Service-Thread");
     }
 
     @Override
     public void run() {
       while (!Thread.interrupted()) {
-        for (Long lastHeartbeat: currentVersionLastHeartbeat.values()) {
-          changeCaptureStats.recordLag(System.currentTimeMillis() - lastHeartbeat);
-        }
         try {
+          recordStats(currentVersionLastHeartbeat, changeCaptureStats, pubSubConsumer.getAssignment());
           TimeUnit.SECONDS.sleep(60L);
         } catch (InterruptedException e) {
-          // We've received an interrupt which is to be expected, so we'll just leave the loop and log
+          LOGGER.warn("Lag Monitoring thread interrupted!  Shutting down...", e);
           break;
         }
       }
-      LOGGER.info("Lag Monitoring thread interrupted!  Shutting down...");
+    }
+
+    protected void recordStats(
+        Map<Integer, Long> currentVersionLastHeartbeat,
+        BasicConsumerStats changeCaptureStats,
+        Set<PubSubTopicPartition> assignment) {
+      for (Map.Entry<Integer, Long> lastHeartbeat: currentVersionLastHeartbeat.entrySet()) {
```

Comment:
It might not be thread-safe here we iterate over Hashmap: `currentVersionLastHeartbeat`. The main thread may modify it. Besides, we expect 
`maxLagSensor = registerSensor("max_partition_lag", new Max());`
to show maximum on dashboard?

Reply:
Entryset should snapshot the entries at time of iteration